lib/spack/spack/solver/concretize.lp

```diff
@@ -373,6 +397,14 @@ attr("node_flag_source", node(X, A1), A2, node(Y, A3))
      imposed_constraint(ID, "node_flag_source", A1, A2, A3),
      condition_set(node(Y, A3), node(X, A1)).
 
+attr("provider_set", node(PDupID, Provider), node(VDupID, Virtual))
+  :- impose(ID, RequestingNode),
+     imposed_constraint(ID, "provider_set", Provider, Virtual),
+     condition_set(RequestingNode, node(PDupID, Provider)),
+     condition_set(RequestingNode, node(VDupID, Virtual)).
```

Comment:
I suspect this rule is what increased the solution time the most.

Reply:
Restricted the scope of this rule in https://github.com/spack/spack/pull/40173/commits/b10d54b5e4ef96a4650db9a559ad7721f9ac93a8